src/mbgl/layout/pattern_layout.hpp

```diff
@@ -90,12 +96,27 @@ class PatternLayout : public Layout {
                     }
                 }
             }
-            features.push_back({static_cast<uint32_t>(i), std::move(feature), patternDependencyMap});
+
+            features.push_back({static_cast<uint32_t>(i), std::move(feature), patternDependencyMap, 0.});
```

Comment:
could we evaluate the sort key already here?

Reply:
in that case we could keep `features` always sorted, the easiest way using of `std::set<PatternFeature>` instead of `std::vector`

Reply:
Maybe I don't know enough about `std::set`, but "being less than" seems to be the requirement for being unique. Because two separate `PatternFeature` may share a sort-key, wouldn't we end up with only one of them recorded in the `std::set`? I would need to come up with some other field to add to the comparison that would ensure all features are seen as "unique"?

Reply:
Also curious, what's the desire to do incremental update of the sort? This is so there's no intermediate state where the features are unsorted?

Reply:
If sort keys are not unique, we could use `std::list` + `std::lower_bound`

> Also curious, what's the desire to do incremental update of the sort? This is so there's no intermediate state where the features are unsorted?

Composing sorted list from the beginning is faster than compose it unsorted and then sort

Reply:
Huh, I didn't know that. I thought they were equivalent complexity and sorting in a batch was usually faster in practice because of reduced relocations and cache efficiency.

I thought about using `std::list` instead of `std::vec` so the features could be inserted incrementally without triggering relocations, but my memory was that `vec` perf in time and memory tends to be much better so I left it and used the batch sort. Happy to implement this with `std::list` though.